cmd/drain_node.go

```diff
@@ -0,0 +1,190 @@
+be removed from a node the following requirements should be met:
+- volume robustness field is "healthy".
+
+Drain node steps:
+1. Disable scheduling on the node.
+2. Iterate through the replicas that live on the node and delete in case the
+above conditions are met, else wait and retry.
+
+In case an error occurs while deleting a replica the command logs the error and
+proceeds, ignoring if the replica is still on the node or not. If this command
+is used to perform upgrades, longhorn will clean any failed leftover replicas
+after a while.
+
+There is no timeout in the command, so if the conditions never allow deleting a
+replica the command will loop forever trying.
+
+example:
+# lhctl drain node [node-id]
+`,
+	PreRun: func(cmd *cobra.Command, args []string) {
+		InitManagerClient()
+	},
+	Run: func(cmd *cobra.Command, args []string) {
+		node, err := parseDrainNodeArgs(cmd, args)
+		if err != nil {
+			eh.ExitOnError(err)
+		}
+
+		if err := validateDrainNode(node); err != nil {
+			eh.ExitOnError(err)
+		}
+
+		// Call disable node from the respective command disableCmd
+		if err := disableNode(node); err != nil {
+			eh.ExitOnError(err)
+		}
+		if err := waitForDisabledNode(node, 10); err != nil {
+			eh.ExitOnError(err)
+		}
+
+		// Drain the node from replicas
+		if err := drainNode(node); err != nil {
+			eh.ExitOnError(err)
+		}
+	},
+}
+
+func init() {
+	drainCmd.AddCommand(drainNodeCmd)
+}
+
+func parseDrainNodeArgs(cmd *cobra.Command, args []string) (string, error) {
+	if len(args) < 1 {
+		cmd.Help()
+		return "", errors.New("No node name specified")
+	}
+
+	node := args[0]
+
+	if len(args) > 1 {
+		fmt.Println(fmt.Sprintf(
+			"ignoring arguments: %v",
+			args[1:],
+		))
+	}
+
+	return node, nil
+}
+
+func validateDrainNode(node string) error {
+	_, err := mc.GetNode(node)
+	return err
+}
+
+func drainNode(nodeName string) error {
+
+	volumes, err := mc.ListVolumes()
+	if err != nil {
+		return err
+	}
+
+	filteredVolumes := filterVolumes(volumes, nodeName)
+	if len(filteredVolumes) == 0 {
+		fmt.Println("Nothing to drain")
+		return nil
+	}
+
+	var wg sync.WaitGroup
+
+	// Iterate through replica list and drain on best effort. If we receive
+	// an error while issuing the remove command or waiting to verify
+	// deletion just log and proceed.
+	for _, volume := range filteredVolumes {
+		for _, replica := range volume.Replicas {
+			if replica.HostId == nodeName {
+				wg.Add(1)
+				go func(replicaName, volumeName string) {
+					defer wg.Done()
+					for {
+						err := drainReplicaFromVolume(
+							replicaName,
+							volumeName,
+						)
+						if err == nil {
+							fmt.Println(fmt.Sprintf(
+								"%v deleted",
+								replicaName,
+							))
+							return
+						}
+						fmt.Println(fmt.Sprintf(
+							"error %v deleting replica: %s, retrying",
+							err,
+							replicaName,
+						))
+						// sleep before retrying
+						time.Sleep(time.Duration(1 * time.Second))
```

Comment:
Perhaps a timeout here?

Reply:
timeout is a good idea. I also think that it should be configurable via a flag on the command